modules/loader/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/Main.scala

```diff
@@ -12,78 +12,74 @@
+import com.snowplowanalytics.snowplow.rdbloader.discovery.DataDiscovery
+import com.snowplowanalytics.snowplow.rdbloader.loading.Common.load
 import com.snowplowanalytics.snowplow.rdbloader.utils.SSH
 
+import io.sentry.Sentry
+
 object Main extends IOApp {
-  /**
-   * If arguments or config is invalid exit with 1
-   * and print errors to EMR stdout
-   * If arguments and config are valid, but loading failed
-   * print message to `track` bucket
-   */
+
   def run(argv: List[String]): IO[ExitCode] =
     CliConfig.parse(argv) match {
-      case Valid(config) =>
-        RealWorld.initialize[IO](config).flatMap { dsls =>
-          import dsls._
-          workStream(config, dsls)
-            .compile
-            .drain
-            .value
-            .attempt
-            .map {
-              case Left(e) =>
-                e.printStackTrace(System.out)
-                (LoaderError.LoaderLocalError(e.getMessage): LoaderError).asLeft
-              case Right(e) => e
-            }
-            .flatMap(res => close[IO](config.logKey, res))
+      case Valid(cli) =>
+        Environment.initialize[IO](cli).flatMap { env =>
+          env.loggingF.info(s"RDB Loader [${cli.config.name}] has started. Listening ${cli.config.messageQueue}") *>
+            process(cli, env)
+              .compile
+              .drain
+              .attempt
+              .flatMap {
+                case Left(e) =>
+                  Sentry.captureException(e)
+                  e.printStackTrace(System.out)
+                  env.loggingF.track(LoaderError.RuntimeError(e.getMessage).asLeft).as(ExitCode.Error)
+                case Right(_) =>
+                  IO.pure(ExitCode.Success)
+              }
         }
       case Invalid(errors) =>
-        IO.delay(println("Configuration error")) *>
-          errors.traverse_(message => IO.delay(println(message))).as(ExitCode.Error)
+        IO.delay(System.err.println("Configuration error")) *>
+          errors.traverse_(message => IO.delay(System.err.println(message))).as(ExitCode(2))
     }
 
-  def workStream(config: CliConfig, dsls: RealWorld[IO]): Stream[LoaderAction[IO, ?], Unit] = {
-    import dsls._
+  /**
+   * Main application workflow, responsible for discovering new data via message queue
+   * and processing this data with loaders
+   *
+   * @param cli whole app configuration
+   * @param env initialised environment containing resources and effect interpreters
+   * @return endless stream waiting for messages
+   */
+  def process(cli: CliConfig, env: Environment[IO]): Stream[IO, Unit] = {
+    import env._
 
-    discover[IO](config).evalMap { case (discovery, _) =>
-      val jdbc = SSH.resource[IO](config.target.sshTunnel) *>
-        JDBC.interpreter[IO](config.target, config.dryRun)
+    DataDiscovery.discover[IO](cli.config, env.state)
+      .pauseWhen[IO](env.isBusy)
+      .evalMap { message =>
+        val jdbc: Resource[IO, JDBC[IO]] = env.makeBusy *>
```

Comment:
No, mostly to prevent parallel loading. In a very best case it will slow down Redshift significantly, in worst case will corrupt data modeling (some pipelines have a requirement to load data within chronological ordering).